### PR TITLE
C cuda

### PR DESCRIPTION
**Pull request by ArrogantGao**
I write the kernel function in C Cuda language in a much simpler way, the performance of this easy GEMM is about 80% of cuBLAS, and then performance of this Tropical version for FP32 is about 60% of cuBLAS (~13 TFLOPS).
Since this version is simpler, I added support for Int32 (60%), Int64 (20%), Float64 (20%) and Bool types.
Other tropical rules are also added, including `max_mul!`, `min_add!` and `and_or!`.

In this pull request, the `.cu` files are stored in the `./deps` dir, and can be built via command `] build`, and I am trying to upload them using `BinaryBuilder.jl` (see https://github.com/JuliaPackaging/Yggdrasil/pull/7341).

### CODE REVIEW

**Comment by GiggleLiu** [comment]
Great work! I have the following test report:
```
Test Summary:            | Pass  Fail  Total   Time
CuTropicalGEMM.jl        | 1107    93   1200  26.7s
  Testing Matrix Max Add |  400          400   9.4s
  Testing Matrix max mul |  307    93    400   7.1s
    Float32 max mul      |    8    92    100   0.9s
    Float64 max mul      |  100          100   4.5s
    Int32 max mul        |   99     1    100   0.5s
    Int64 max mul        |  100          100   1.1s
  Testing Matrix Min Add |  400          400   6.3s
```

Once these tests can be fixed, we can merge it into the master branch.
Besides, I left an issue here #7 , which can be a future direction to move.

File: test/test_minadd.jl
Comment (GiggleLiu) [inline test/test_minadd.jl:62]: It is better to test on matrices with designed size rather than random size to cover more corner cases. For example,
```julia
M, N, K = 0, 0, 0
M, N, K = 2, 0, 2
...
```

**Review by GiggleLiu — CHANGES_REQUESTED** [review]
Good job! after fixing the tests, we can get this PR merged.

**Comment by GiggleLiu** [comment]
@ArrogantGao Nice work. Please resolve the conflicts, otherwise I can not merge the PR. If possible, please also get special sized (like size 0 and 1) matrices also tested, which should be straight forward to implement.

File: test/test_minadd.jl
Comment (GiggleLiu) [reply test/test_minadd.jl]: Here, I want to emphasis that we should get zero and one sized matrices tested.

**Comment by ArrogantGao** [comment]
Ops, sorry about the added changes. I forget that they will be uploaded.
I fixed the mistake in the boundary condition, which will make the result unstable when the matrices are small (https://github.com/Cjkkkk/CUDA_gemm/issues/6), so that now the `mutadd` and `maxadd` functions are correct.
I will wrap the rest of the code and add the tests needed tomorrow.

**Comment by ArrogantGao** [comment]
The changes are uploaded, and tests passed on my mechina
```
     Testing Running tests...
Test Summary:     | Pass  Total     Time
CuTropicalGEMM.jl |   63     63  1m38.2s
     Testing CuTropicalGEMM tests passed 
```

There are a few changes:
1. The C-CUDA code provide api for `mul-add`, `max-mul`, `max-add` and `and-or` for row-major matrices. Then in julia, I wrap them as `mul-add`, `max-mul`, `min-mul` , `max-add`, `min-add` and `and-or` for col-major matrices using `permutedims`, since that is faster than directly use C-CUDA code for col-major matrices.
2. The test code have be change, I use `for (M, N, K) in [(5, 6, 7), (101, 102, 103), (1024, 1024, 1024)]` to test different cases. However, it seems that `permutedims` does not work for $0d$ and $1d$ matrices, it that necessary?
3. A benchmark folder is added, which can directly benchmark the C-CUDA code, the results are shown below.

```
Benchmarking AndOrBool. 
benchmarking by large matrix 4096, 4096, 4096
SGEMM with dynamic Matrix size: 1.600392 TFLOPS
time per iteration: 85.878288 ms
 
Benchmarking MaxAddFP32. 
check_all by small matrix 125, 46, 37
Matrix_C check all: OK
benchmarking by large matrix 4096, 4096, 4096
SGEMM with dynamic Matrix size: 13.209873 TFLOPS
time per iteration: 10.404261 ms
 
Benchmarking MaxAddFP64. 
check_all by small matrix 213, 21, 102
Matrix_C check all: OK
benchmarking by large matrix 4096, 4096, 4096
SGEMM with dynamic Matrix size: 2.875586 TFLOPS
time per iteration: 47.795116 ms
 
Benchmarking MaxAddInt32. 
check_all by small matrix 213, 21, 102
Matrix_C check all: OK
benchmarking by large matrix 4096, 4096, 4096
SGEMM with dynamic Matrix size: 14.068782 TFLOPS
time per iteration: 9.769073 ms
 
Benchmarking MaxAddInt64. 
check_all by small matrix 213, 21, 102
Matrix_C check all: OK
benchmarking by large matrix 4096, 4096, 4096
SGEMM with dynamic Matrix size: 3.735243 TFLOPS
time per iteration: 36.795189 ms
 
Benchmarking MaxMulFP32. 
check_all by small matrix 213, 21, 102
Matrix_C check all: OK
benchmarking by large matrix 4096, 4096, 4096
SGEMM with dynamic Matrix size: 13.003628 TFLOPS
time per iteration: 10.569278 ms
 
Benchmarking MutAddFP32. 
check_all by small matrix 5, 6, 7
Matrix_C check all: OK
benchmarking by large matrix 4096, 4096, 4096
SGEMM with dynamic Matrix size: 15.014899 TFLOPS
time per iteration: 9.153505 ms
 
Benchmarking MutAddFP64. 
check_all by small matrix 5, 6, 7
Matrix_C check all: OK
benchmarking by large matrix 4096, 4096, 4096
SGEMM with dynamic Matrix size: 7.832789 TFLOPS
time per iteration: 17.546619 ms
 
Benchmarking MutAddInt32. 
check_all by small matrix 213, 21, 102
Matrix_C check all: OK
benchmarking by large matrix 4096, 4096, 4096
SGEMM with dynamic Matrix size: 15.195343 TFLOPS
time per iteration: 9.044807 ms
 
Benchmarking MutAddInt64. 
check_all by small matrix 213, 21, 102
Matrix_C check all: OK
benchmarking by large matrix 4096, 4096, 4096
SGEMM with dynamic Matrix size: 4.136473 TFLOPS
time per iteration: 33.226120 ms
```

**Comment by ArrogantGao** [comment]
> Great work! I have the following test report:
> 
> ```
> Test Summary:            | Pass  Fail  Total   Time
> CuTropicalGEMM.jl        | 1107    93   1200  26.7s
>   Testing Matrix Max Add |  400          400   9.4s
>   Testing Matrix max mul |  307    93    400   7.1s
>     Float32 max mul      |    8    92    100   0.9s
>     Float64 max mul      |  100          100   4.5s
>     Int32 max mul        |   99     1    100   0.5s
>     Int64 max mul        |  100          100   1.1s
>   Testing Matrix Min Add |  400          400   6.3s
> ```
> 
> Once these tests can be fixed, we can merge it into the master branch. Besides, I left an issue here #7 , which can be a future direction to move.

I will have a check of that later.

**Comment by GiggleLiu** [comment]
Some tests do not pass on my local machine:
```julia
    Testing Running tests...
(i, j, sum, D[i, j]) = (4, 2, -490, 0)
Int32 max add: Test Failed at /home/leo/.julia/dev/CuTropicalGEMM/test/test_maxadd.jl:85
  Expression: check_all(A, B, C, D)

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.0-beta2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] macro expansion
   @ ~/.julia/dev/CuTropicalGEMM/test/test_maxadd.jl:85 [inlined]
 [3] macro expansion
   @ ~/.julia/juliaup/julia-1.10.0-beta2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [4] macro expansion
   @ ~/.julia/dev/CuTropicalGEMM/test/test_maxadd.jl:73 [inlined]
 [5] macro expansion
   @ ~/.julia/juliaup/julia-1.10.0-beta2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [6] top-level scope
   @ ~/.julia/dev/CuTropicalGEMM/test/test_maxadd.jl:39
Test Summary:            | Pass  Fail  Total     Time
CuTropicalGEMM.jl        |   62     1     63  1m11.3s
  Testing Matrix mul add |   12           12    15.7s
  Testing Matrix Max Add |   11     1     12    10.1s
    Float32 max add      |    1            1     0.0s
    Float64 max add      |    1            1     0.0s
    Int32 max add        |          1      1     0.3s
    Int64 max add        |    1            1     0.0s
    Float32 max add      |    1            1     0.0s
    Float64 max add      |    1            1     0.0s
    Int32 max add        |    1            1     0.0s
    Int64 max add        |    1            1     0.0s
    Float32 max add      |    1            1     2.7s
    Float64 max add      |    1            1     3.1s
    Int32 max add        |    1            1     1.9s
    Int64 max add        |    1            1     2.2s
  Testing Matrix Min Add |   12           12    10.0s
  Testing Matrix max mul |   12           12     9.4s
  Testing Matrix min mul |   12           12    10.4s
  Testing Matrix And Or  |    3            3     0.5s
```

Another run shows a different error
```julia
     Testing Running tests...
(i, j, sum, D[i, j]) = (3, 3, 19, 0)
Int32 min add: Test Failed at /home/leo/.julia/dev/CuTropicalGEMM/test/test_minadd.jl:84
  Expression: check_all_minadd(A, B, C, D)

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.0-beta2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] macro expansion
   @ ~/.julia/dev/CuTropicalGEMM/test/test_minadd.jl:84 [inlined]
 [3] macro expansion
   @ ~/.julia/juliaup/julia-1.10.0-beta2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [4] macro expansion
   @ ~/.julia/dev/CuTropicalGEMM/test/test_minadd.jl:72 [inlined]
 [5] macro expansion
   @ ~/.julia/juliaup/julia-1.10.0-beta2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [6] top-level scope
   @ ~/.julia/dev/CuTropicalGEMM/test/test_minadd.jl:38
Test Summary:            | Pass  Fail  Total     Time
CuTropicalGEMM.jl        |   62     1     63  1m13.3s
  Testing Matrix mul add |   12           12    12.5s
  Testing Matrix Max Add |   12           12    13.0s
  Testing Matrix Min Add |   11     1     12    13.1s
    Float32 min add      |    1            1     0.3s
    Float64 min add      |    1            1     0.1s
    Int32 min add        |          1      1     0.4s
    Int64 min add        |    1            1     0.1s
    Float32 min add      |    1            1     0.0s
    Float64 min add      |    1            1     0.0s
    Int32 min add        |    1            1     0.0s
    Int64 min add        |    1            1     0.0s
    Float32 min add      |    1            1     2.5s
    Float64 min add      |    1            1     4.6s
    Int32 min add        |    1            1     1.8s
    Int64 min add        |    1            1     3.3s
  Testing Matrix max mul |   12           12     9.9s
  Testing Matrix min mul |   12           12     9.4s
  Testing Matrix And Or  |    3            3     0.6s
ERROR: LoadError: Some tests did not pass: 62 passed, 1 failed, 0 errored, 0 broken.
in expression starting at /home/leo/.julia/dev/CuTropicalGEMM/test/runtests.jl:5
ERROR: Package CuTropicalGEMM errored during testing
```
These tests seems to be unstable, do you have any idea why this test fails?

**Comment by ArrogantGao** [comment]
It seems that in both cases the problem is about the `maxadd_Int32` kernel (`minadd` is based on `maxadd`).
A possible reason is that for matrices that cannot be evenly divided, I used zero padding technique
For the `max-add` rule, the proper padding number should be `-Inf` but I used `INT_MIN` instead for `Int` kernel, since `Int` type do not have `Inf`.
So maybe it is because there are some kind of overflow and leads to the $0$ shown in error message?

However, I did not found the error in tests on the server, both in `julia` or `C`.
Currently, I have no idea about how to solve that, so I disabled the `maxadd_Int` functions.

Another update is that I wrapped all function as `matmul!(A, B, C, kernel)`, and it support $0d$ and $1d$ matrices now. The tests are done over `for (M, N, K) in [(0, 0, 0), (2, 0, 0), (2, 2, 0), (5, 6, 7), (101, 102, 103), (1024, 1024, 1024)]`.

The new tests passed on the server:
```
Test Summary:     | Pass  Total     Time
CuTropicalGEMM.jl |  102    102  1m20.6s
     Testing CuTropicalGEMM tests passed 
```

**Comment by GiggleLiu** [comment]
For integer types, I used to use the following dirty trick:
https://github.com/TensorBFS/TropicalNumbers.jl/blob/6200c3bd992a13d6dba193f86279246a98fb0513/src/tropical.jl#L6C32-L6C32

Because integer min can easily overflow.

**Comment by GiggleLiu** [comment]
Now all tests pass, if you could resolve the confilicts with main, we can get this PR merge immediately. Good job!

**Comment by ArrogantGao** [comment]
> For integer types, I have used the following dirty trick: https://github.com/TensorBFS/TropicalNumbers.jl/blob/6200c3bd992a13d6dba193f86279246a98fb0513/src/tropical.jl#L6C32-L6C32
> 
> Because integer min can easily overflow. e.g.
> 
> ```julia
> julia> typemin(Int) + -3
> 9223372036854775805
> ```

Indeed, I will try some finite lower bound later, maybe `INT_MIN/2` will work for matrices that satisify`INT_MIN/2 < x < INT_MAX/2`, since the rule is `max-add`.

**Comment by ArrogantGao** [comment]
> > > For integer types, I have used the following dirty trick: https://github.com/TensorBFS/TropicalNumbers.jl/blob/6200c3bd992a13d6dba193f86279246a98fb0513/src/tropical.jl#L6C32-L6C32
> > > Because integer min can easily overflow. e.g.
> > > ```julia
> > > julia> typemin(Int) + -3
> > > 9223372036854775805
> > > ```
> > 
> > 
> > Indeed, I will try some finite lower bound later, maybe `INT_MIN/2` will work for matrices that satisify`INT_MIN/2 < x < INT_MAX/2`, since the rule is `max-add`.
> 
> Unfortunately, `INT_MIN/2` does not work in practise. Since by adding two tropical zeros, the result can still overflow. Maybe try `sqrt(INT_MIN)`?

`sqrt(INT_MIN)` is too small, and is already covered by `Float32`
```
julia> typemax(Int32)
2147483647

julia> sqrt(typemax(Int32))
46340.950001051984

julia> b = 46340f0
46340.0f0
```

A possible approach is setting the Tropical zero as `t0 = INT_MIN / 2 + 1`, and make $x \in [t_0 / 2, t_0 / 2]$.
Then the $t_0 + t_0 = INT_MIN + 2 > INTMIN $, and for any $x, y$, we have $max(x, y + t_0) = x \leq t_0$.

But I think in that case there seems to be no advantages using `Int` instead of `Float` directly.

**Pull request merged**
